src/utils/aadV2TokenProvider.ts

```diff
@@ -0,0 +1,278 @@
+import { Clipboard, commands, env, Uri, window } from 'vscode';
+import * as Constants from '../common/constants';
+import { HttpRequest } from '../models/httpRequest';
+import { HttpClient } from './httpClient';
+import { EnvironmentVariableProvider } from './httpVariableProviders/environmentVariableProvider';
+
+/*
+ AppId provisioned to allow users to explicitly consent to permissions that this app can call
+*/
+export const AadV2TokenProviderClientId = "07f0a107-95c1-41ad-8f13-912eab68b93f";
+
+export class AadV2TokenProvider {
+    private readonly _httpClient: HttpClient;
+    private readonly clipboard: Clipboard;
+
+    public constructor() {
+        this._httpClient = new HttpClient();
+        this.clipboard = env.clipboard;
+    }
+
+    public async AcquireToken(name : string): Promise<string> {
+
+        const authParams = new AuthParameters();
+        await authParams.ParseName(name);
+
+        if (!authParams.forceNewToken) {
+            const tokenEntry = AadV2TokenCache.GetToken(authParams.GetCacheKey());
+            if (tokenEntry && tokenEntry.SupportScopes(authParams.scopes)) {
+                return tokenEntry.Token;
+            }
+        }
+
+        if (authParams.appOnly) {
+            return await this.GetConfidentialClientToken(authParams);
+        }
+
+        const deviceCodeResponse: IDeviceCodeResponse = await this.GetDeviceCodeResponse(authParams);
+        const isDone = await this.promptForUserCode(deviceCodeResponse);
+        if (isDone) {
+            return await this.GetToken(deviceCodeResponse, authParams);
+        } else {
+            return "";
+        }
+    }
+
+    private async GetDeviceCodeResponse(authParams: AuthParameters) : Promise<IDeviceCodeResponse> {
+        const request = this.createUserCodeRequest(authParams.clientId, authParams.tenantId, authParams.scopes);
+        const response = await this._httpClient.send(request);
+
+        const bodyObject = JSON.parse(response.body);
+
+        if (response.statusCode !== 200) {
+            // Fail
+            this.processAuthError(bodyObject);
+        }
+
+        if (bodyObject.error) {  // really!?
+            this.processAuthError(bodyObject);
```

Comment:
200 status code will also contain error response, if so, can we judge the error state from the `error` only, or merge with line 52

Reply:
The 200 response is an error from the AAD V2 device code endpoint.  An issue has been filed with service to get it fixed.  HTTP status codes should always be checked for error conditions.  Once the error is fixed we will be able to remove this extra check.

Reply:
I updated the comment to be more clear.